### PR TITLE
Pre fetch data, save in a model

### LINKED ISSUE #202: Pre-calculate eligiblity
As a user of the system, i want the pages to load in a reasonable amount of time, which can be a reality if we precalculate eligibility, so we can display it straight out of the database

### PR DESCRIPTION
**Pull request by Br3nda**
I've made a rake task to fetch the data (seven days at a time, but the caller can choose how many)
I added a model (eligibility) to save the results into.

I've changed the controller to read the eligiblities table, but to fall back to calling the service if there's no data yet. When it does fall back on that it only retrieves one day

I modified the controller client#eligibility method to return a hash, which should be the same as #211 expects too

fixes #202

### CODE REVIEW

File: app/services/eligibility_service.rb [hunk 1]
```diff
@@ -1,90 +1,99 @@
...
     @response['persons'][person_name][key]
   end
 
+  private
+
   def query
+    all_persons = {}
+    all_names = []
+    (0..@number_of_days).each do |n|
+      day = (Date.parse(@day) + n).strftime('%Y-%m-%d')
```
Comment (danieljauch) [inline app/services/eligibility_service.rb]: This is a repeat from line 16, it should probably be pulled out into its own method.

File: app/services/eligibility_service.rb [hunk 2]
```diff
@@ -1,90 +1,99 @@
...
+      @eligibility.each_year_presence = person_result(person_name, 'citizenship__meets_each_year_minimum_presence_requirements')[day]
+
+      # hash for each rolling year of five
+      @eligibility.present_days_by_rolling_year = person_result person_name, 'days_present_in_new_zealand_in_preceeding_year'
+      @eligibility.mimimum_presence_by_rolling_year = person_result person_name, 'citizenship__meets_preceeding_single_year_minimum_presence_requirement'
+
+      @eligibility.save!
+
+      Rails.logger.info @eligibility
+    end
```
Comment (danieljauch) [inline app/services/eligibility_service.rb]: This method is starting to get long and complex. Are there any ways we can abstract some of this functionality?

File: app/services/eligibility_service.rb [hunk 3]
```diff
@@ -101,20 +110,18 @@ def presence_values
...
-  def years_before(num_years)
-    (@day.to_date - num_years.year).strftime(date_format)
+  def years_before(day, num_years)
+    (day.to_date - num_years.year).strftime(EligibilityService.date_format)
   end
 
-  def date_format
+  def self.date_format
     '%Y-%m-%d'
   end
```
Comment (danieljauch) [inline app/services/eligibility_service.rb]: Is this used anymore? We could probably use it on that duplicated line.

File: app/services/eligibility_service.rb [hunk 4]
```diff
@@ -101,20 +110,18 @@ def presence_values
...
 
-  def person_name
-    'Ruby'
-  end
-
   def calculate(query)
+    Rails.logger.info 'Calculating'
     Rails.cache.fetch(query) do
+      Rails.logger.info 'Requesting from OpenFisca'
```
Comment (danieljauch) [inline app/services/eligibility_service.rb]: Are you meaning to keep all of these logs in the server? There are quite a few throughout this PR.

File: app/services/eligibility_service.rb (hunk 3 above)
Comment (Br3nda) [reply app/services/eligibility_service.rb]: Good catch. I've DRY-ed this.

File: app/services/eligibility_service.rb (hunk 2 above)
Comment (Br3nda) [reply app/services/eligibility_service.rb]: i think it should be it's own class - or part of the eligibility model

Comment (Br3nda) [reply app/services/eligibility_service.rb]: i've moved it

**Pull request merged**